messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt

```diff
@@ -0,0 +1,88 @@
+package io.customer.messagingpush.processor
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import io.customer.messagingpush.MessagingPushModuleConfig
+import io.customer.messagingpush.util.PushTrackingUtil
+import io.customer.sdk.data.request.MetricEvent
+import io.customer.sdk.repository.TrackRepository
+import io.customer.sdk.util.Logger
+
+internal class PushMessageProcessorImpl(
```

Comment:
I am not sure if I find this class very useful at the moment. It seems to contain mostly Queue logic for the `PushMessageProcessor.recentMessagesQueue` data structure and a small bit of code beyond that. 

I am leaning towards removing it. 

For the Queue data structure, if we were to [use a different data type for the `PushMessageProcessor.recentMessagesQueue` data structure](https://github.com/customerio/customerio-android/pull/204/files#r1204632825) or we create our own thread-safe, maximum size, Queue data structure class. This would remove a lot of the need for the `getOrUpdateMessageAlreadyProcessed` function. 

As for the rest of the code that is specific to FCM and GCM push processing, I would rather have the code contained inside of the existing FCM Service code and the GCM Broadcast Receiver. 

Having a separate class that only contains *mostly* this core code: 

```kt
if (!deliveryMetricsQueue.processedMessageAlready(deliveryId = deliveryId)) {
  if (moduleConfig.autoTrackPushEvents) {
     trackRepository.trackMetric(...)
  }
}
```

Seems not very useful to me at the moment.

Reply:
Our current queue structure doesn't have any support for preventing duplications, the processor class not only takes the responsibility for avoiding duplication but also makes it easier for other classes. With the introduction of GCM receiver, we'll now be doing several push related operations at multiple places. So there has to be one class which holds all push related operations at a single place and avoid code repetition. The processor class takes this responsibility and will be used for adding more operations in future as well.